employee-rostering-server/src/main/java/org/optaweb/employeerostering/server/util/HierarchyTree.java

```diff
@@ -0,0 +1,192 @@
+/*
+ * Copyright 2019 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.optaweb.employeerostering.server.util;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+/**
+ * A data structure representing a hierarchy tree. Each node in the tree represent
+ * a subclass of the Hierarchy. Can be used to find the most specific class
+ * an object is applicable to (example: in a Hierarchy tree containing Throwable and RuntimeException,
+ * IllegalStateException should map to RuntimeException, not Throwable).
+ * @param <K> The type of the key element in the Hierarchy
+ * @param <V> The type of the value element in the Hierarchy
+ */
+public class HierarchyTree<K, V> {
```

Comment:
@Christopher-Chianelli @ge0ffrey was it the plan to fully implement https://issues.jboss.org/browse/PLANNER-933 now? Given we want to rewrite most of the application, we should think twice about adding every such a complex piece of code.

Reply:
I think the rewrite apply mostly to the client code (i.e. most of the server code will remain the same). `HierarchyTree` is server code, so the probability of it changing in the rewrite is low.